lib/FSi/DoctrineExtensions/Translatable/TranslatableListener.php

```diff
@@ -342,250 +311,68 @@ private function updateObjectTranslations(ObjectManager $objectManager, $object)
-     */
-    private function findOrCreateObjectTranslation(ObjectManager $objectManager, $object, $translationAssociation, $locale)
-    {
-        $translation = $this->getRepository($objectManager, $object)
-            ->getTranslation($object, $locale, $translationAssociation);
+            $hasTranslatedProperties = $this->translationHelper->hasTranslatedProperties($context, $object);
 
-        if (!$objectManager->contains($translation)) {
-            $objectManager->persist($translation);
-        }
-
-        return $translation;
-    }
-
-    /**
-     * @param \Doctrine\Common\Persistence\ObjectManager $objectManager
-     * @param object $object
-     * @throws Exception\AnnotationException
-     * @return TranslatableRepository
-     */
-    private function getRepository(ObjectManager $objectManager, $object)
-    {
-        $meta = $this->getObjectClassMetadata($objectManager, $object);
-        $repository = $objectManager->getRepository($meta->getName());
-
-        if (!($repository instanceof TranslatableRepositoryInterface)) {
-            throw new Exception\AnnotationException(sprintf(
-                'Entity "%s" has "%s" as its "repositoryClass" which does not implement \FSi\DoctrineExtensions\Translatable\Model\TranslatableRepositoryInterface',
-                $meta->getName(),
-                get_class($repository)
-            ));
-        }
-
-        return $repository;
-    }
+            if (!isset($locale) && $hasTranslatedProperties) {
+                throw new Exception\RuntimeException(
+                    "Neither object's locale nor the current locale was set for translatable properties"
+                );
+            }
 
-    /**
-     * @param \Doctrine\Common\Persistence\ObjectManager $objectManager
-     * @param object $object
-     * @param string $translationAssociation
-     * @return array
-     */
-    private function clearObjectProperties(ObjectManager $objectManager, $object, $translationAssociation)
-    {
-        $translatableProperties = $this->getTranslatableMetadata($objectManager, $object)->getTranslatableProperties();
-        $translationMeta = $this->getTranslationClassMetadata($objectManager, $object, $translationAssociation);
-        foreach ($translatableProperties[$translationAssociation] as $property => $translationField) {
-            if ($translationMeta->isCollectionValuedAssociation($translationField)) {
-                $this->getPropertyAccessor()->setValue($object, $property, array());
+            if ($hasTranslatedProperties) {
+                $this->translationHelper->copyPropertiesToTranslation(
+                    $objectManager,
+                    $context,
+                    $object,
+                    $locale
+                );
             } else {
-                $this->getPropertyAccessor()->setValue($object, $property, null);
+                $this->translationHelper->removeEmptyTranslation(
+                    $objectManager,
```

Comment:
`$objectManager` get be retrieved from `$context` and does not need to be passed directly